components/script/dom/htmlelement.rs

```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
             None => false,
         }
     }
+
+    /// <https://html.spec.whatwg.org/multipage/#rendered-text-fragment>
+    fn rendered_text_fragment(&self, input: DOMString) -> DomRoot<DocumentFragment> {
+        // Step 1: Let fragment be a new DocumentFragment whose node document is document.
+        let document = document_from_node(self);
+        let fragment = DocumentFragment::new(&document);
+
+        // Step 2: Let position be a position variable for input, initially pointing at the start
+        // of input.
+        let mut position = input.chars().peekable();
+
+        // Step : Let text be the empty string.
+        let mut text = String::new();
+
+        // Step 4
+        while let Some(ch) = position.next() {
+            match ch {
+                // While position is not past the end of input, and the code point at position is
+                // either U+000A LF or U+000D CR:
+                '\u{000A}' | '\u{000D}' => {
+                    if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
+                        // a \r\n pair should only generate one <br>,
+                        // so just skip the \r.
+                        position.next();
+                    }
+
+                    if !text.is_empty() {
+                        append_text_node_to_fragment(&document, &fragment, text);
+                        text = String::new();
+                    }
+
+                    let br = HTMLBRElement::new(local_name!("br"), None, &document, None);
+                    fragment.upcast::<Node>().AppendChild(br.upcast()).unwrap();
+                },
+                _ => {
+                    // Collect a sequence of code points that are not U+000A LF or U+000D CR from
+                    // input given position, and set text to the result.
+                    text.push(ch);
+                },
+            }
+        }
+
+        // If text is not the empty string, then append a new Text node whose data is text and node
+        // document is document to fragment.
+        if !text.is_empty() {
```

Comment:
I'm confused about the loop, do you still need this here besides what is at [966](https://github.com/servo/servo/pull/33312/files#diff-978993aad649a5f5ef459e8071ed6b982d038e71ca6d53156e9dcdf74192c702R966)?

Reply:
Not sure whether the latest changes broke the line numbers, do you mean why there's the following code both inside the `while` loop and at the end?
```rust
if !text.is_empty() {
    append_text_node_to_fragment(&document, &fragment, text);
    text = String::new();
}
```

From what I understand that's because linefeeds should be converted to `<br>` tags, which means we might need to generate multiple text nodes. And since we might not end on a linefeed we have to put the accumulated characters into a final text node.

Reply:
Ok, thanks for clarifying.